### ISSUE TITLE
Click in commit message textfield leads to change on selected item from staged/unstaged area

### ISSUE THREAD

**Issue by FireEmerald** [gitextensions/gitextensions#8468]
## Current behaviour
![ezgif-7-1de3e924b723](https://user-images.githubusercontent.com/1107939/93665311-c1eeed00-fa75-11ea-86c0-d37c0f0054a2.gif)

## Expected behaviour
A click inside the textbox which contains the commit message should **not** change the selected file nor the position inside the shown file. _I must note that the bug is horrible for the user experience._

## Steps to reproduce

1. Create a empty repository in windows
2. Add various files, stage at least one file
3. Select a single file from unstage list and click inside the commit message textfield

## Did this work in previous version of GitExtensions
I updated today GitExtensions and think that the issue did not exist in the previous version. I only found one similar issue #4894.

## Environment
- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4220.0
- DPI 96dpi (no scaling)

**Comment by vbjay** [comment]
This is by design. If you are writing the commit message then we want to
make it easy to review the changes to be committed.

**Comment by FireEmerald** [comment]
You don't make it easy. You overcomplicate a user interface with a auto switch to a random file which the user may do not want to select.

If i click inside the commit message i want to write the message and not switch other views. 

Please leave it up to the user how he or she wants to use the application. If the user would like to view the changes to be committed, he can select the file.

**Comment by vbjay** [comment]
Not random. First staged.

She's it make sense to write a commit message for unstaged changes?  Nope.

**Comment by pmiossec** [comment]
Yes. That's on purpose. Because, for me, the workflow is this one:

1. Review each files and stage them
2. Once finished, enter your message describing what have been done. And to do so, you sometimes have to look again what you have staged.
And if you are already focused on the staged file, you can easily see them. And if needed, you can use the shortcut 'ctrl+n/p' to pass to the next or previous file. So that you can review the staged files without exiting the commit message textarea.

**Comment by FireEmerald** [comment]
> Does it make sense to write a commit message for unstaged changes?  Nope.

That's wrong. Selecting a file which is not yet staged does not mean it has no information about the recent commit. Just think about a different workflow where I first select a unstaged file, write about it, stage it and continue with the next unstaged file until all files are staged.

**Comment by vbjay** [comment]
Stage then write and you will be fine. You can even stage specific lines if
needed.

**Comment by FireEmerald** [comment]
> Stage then write and you will be fine.

Think you don't want to understand that there are other users which may use GitExtensions in a different way as you do. There is no "right" way it should be up to the user.

**Comment by vbjay** [comment]
I'll let the team handle it from here. It doesn't make sense to write a
commit message for unstaged changes at all. This is why we did this. It is
on purpose.

**Comment by RussKie** [comment]
Every user may have own preferences as to how they stage/unstage/prepare to commit. I do agree that some of the existing behaviours (though long standing) have questionable UX.
We do have a lot of intricate and complicated logic in the FormCommit that we can probably do away with. That said, this isn't a high priority item for me personally.

@FireEmerald if you feel strongly about this, you're welcome to offer your vision and open a PR, but it will need to sway longstanding contributors it is a worthy change.

**Comment by mstv** [comment]
The current behavior helps (at least me) to not forget to stage the changes. 
I propose to let the the Stage button ![grafik](https://user-images.githubusercontent.com/36601201/94343887-9e024d00-001b-11eb-8b48-82fab158026c.png) and its mnemonic <kbd>Alt</kbd>+<kbd>S</kbd> (but not the hotkey <kbd>S</kbd>) always do additionally:
1. select the staged file in the Stage Files list
2. restore the view position of the file
3. focus the Commit Message textfield

(as it already does if it was the last file to stage).
So one can still stage files one after the other using the hotkey <kbd>S</kbd>.
(If the button is clicked using the mouse, it should not disturb that the Commit Message would be focused then.)

**Comment by RussKie** [comment]
The logic in the `FormCommit` is already beyond comprehension, adding more nuanced flows would only make the problem worse.
In the end if a user wants to do something unusual (or stupid), the app can only do so much to prevent it. The user has to take responsibility too.

**Comment by sylvain-bougnoux** [comment]
I do agree with @FireEmerald, I find in _my process_ this behavior quite awkward.
For instance, as shown implicitly in the animation above, I have a long source with diffs, and I want to stage just some of them. Once I find one I press 'S' over the source area, and continue scrolling down. **From time to time, I build my commit message** with what I staged (**like taking notes**). With this new behavior, as said the source editor switches to selected stages, and I lost where I was. Actually it drives me crazy.
Maybe it could simply be an option in the settings 'Detailed/Commit dialog' ?

**Comment by RandScullard** [comment]
I agree with @FireEmerald that this is a very annoying behavior. Simply clicking in the commit message field should not change the state of any other parts of the commit window! 

@vbjay It is not necessarily true that when writing the commit message we only need to refer to staged changes. If that works for you, that's great, but please don't assume that your workflow is the only workflow.

**Comment by vbjay** [comment]
@RandScullard  You would have a ton more people complaining about hey I'm not seeing the files 8m commiting. It makes perfect sense to be writing commit messages about the files being committed NOT the unstaged files. Why it selects the staged file. Let's you see the first file and can cycle by hitting alt + up/down. It is designed to help review the commit you are about to make and not the unstaged files.

**Comment by RandScullard** [comment]
@vbjay I totally get why it makes perfect sense for your workflow. I'm just asking you to be open to other users' workflows where it doesn't make sense.

**Comment by vbjay** [comment]
Please do show a workflow where it makes sense. The commit message is at bottom and to right for a reason. Since when do you write a commit message about unstaged files?

**Comment by RandScullard** [comment]
I often have a bunch of files to commit, with various changes that I want to specifically call out in the "body" of the commit message. I start with all of the files unstaged. One by one, I click on an unstaged file, review the diff, and decide whether there's anything in there that needs special mention in the commit message. If so, I click in the commit message and add some text. (The text isn't final at this point, I'm just taking notes.) 

Once I'm done with a file, I stage it and move on to the next file. As I go along, I use the list of staged files as the list of files I'm done with - very helpful if I get interrupted and need to come back later! (And I often find I need to jump around in the file list - it's not always top-to-bottom order. At those times it becomes important to have that visual reminder of where I am in the process.)

Once all the files are staged, I go back to the commit message and clean up the notes I've written and format them into something reasonable. Then I commit.

Are there other ways I could do this? Probably, but this works well for me - and I've been using Git Extensions this way for quite a few years now. In general, Git Extensions does a great job of staying out of the way and not trying to impose any particular workflow on the user. That's all I'm asking for here.

**Comment by vbjay** [comment]
Stage the file before taking notes. Then use u in diff view to unstage changes you don't want. Or use s to stage specific lines and then make notes. You won't have any issues it the file is staged even if just one line.

**Comment by RandScullard** [comment]
Here's what happens when I try your suggestion: I stage a file. This automatically selects the next unstaged file. I click in the commit message field. The **first** staged file gets selected, not the file I just staged a second ago. I now have to hunt through the staged file list for the file I just staged and select it again. This is not convenient. Am I missing something?

**Comment by vbjay** [comment]
Now that could be room for improvement. Selecting the file last staged if last staged was only one file if not just select the first like it does.

Also in the meantime use alt up/down to cycle files in staged while in commit message.

**Comment by RandScullard** [comment]
Selecting the last file staged would be an improvement, but I would still want a way to turn off this behavior and be able to click in the commit message freely without any side effects.
